metadata.yaml

```diff
@@ -29,6 +29,9 @@ peers:
     interface: kube-masters
 provides:
   kube-api-endpoint:
+    # kube-api-endpoint is deprecated. Its functionality has been rolled into
+    # the kube-control interface. The relation endpoint will be removed in a
+    # future release.
```

Comment:
kubernetes-e2e and tigera-secure-ee will also need to be updated since they depend on this endpoint:

https://github.com/charmed-kubernetes/charm-kubernetes-e2e/blob/96078da927cda0de6448fbe25ba16c5071966177/metadata.yaml#L21-L22

https://github.com/charmed-kubernetes/layer-tigera-secure-ee/blob/99f4b08d3b3f7e6efcd11aca14150c617bfdb791/metadata.yaml#L27-L28

Reply:
I think we lose something by deprecating kube-api-endpoint. It was possible for users to stand up a reverse-proxy (e.g. apache/nginx) in front of the Kubernetes API. Do we know for sure that users aren't dependent on this?

Reply:
The relation pattern doesn't affect whether you can put a reverse proxy in front of the API server; in fact, that's exactly what the kubapi-load-balancer charm is doing.  It just changes how the information about that information is disseminated between the charms.

For comparison, let me outline the current situation vs this proposal.

## Current

The current setup depends on whether you're using kubeapi-load-balancer or not:

### Without kubeapi-load-balancer

  * k8s-master <-- kube-api-endpoint --> k8s-worker (communicates API endpoint address to workers)
  * k8s-master <-- kube-control --> k8s-worker (communicates all other control plane info to workers)
  * k8s-master <-- lb-provider --> cloud integrator, (maybe) haproxy, etc (optional)

### With kubeapi-load-balancer

  * k8s-master <-- loadbalancer --> kubeapi-load-balancer (communicates the proxy API endpoint address to the master)
  * k8s-master <-- kube-api-endpoint --> kubeapi-load-balancer (communicates the actual API endpoint addresses to the proxy)
  * k8s-master <-- kube-control --> k8s-worker (communicates all the other control plane info to the workers)
  * kubeapi-load-balancer <-- kube-api-endpoint --> k8s-worker (communicates the proxy API endpoint address to the workers)

## Proposal

For all cases it would be:

  * k8s-master <-- kube-control --> k8s-worker (communicates all control plane info, including API endpoint, to the workers)
  * k8s-master <-- lb-provider --> kubeapi-load-balancer, cloud integrator, (maybe) haproxy, etc (optional)

In this case, the kubeapi-load-balancer becomes just another optional LB provider.

Reply:
Right, I get that. My concern is that kubeapi-load-balancer might not be the only reverse proxy people are using. Are we going to add new loadbalancer relations to charms like haproxy and apache2 as well? Just how much work are we creating by making this change?